src/picongpu/include/fields/background/templates/bgrTWTS.hpp

```diff
@@ -0,0 +1,953 @@
+                 * The abs()-function is for correct offset for -phi<-90Deg and +phi>+90Deg. */
+                const float_64 y1=float_64(halfSimSize[0]
+                                    *picongpu::SI::CELL_WIDTH_SI)*abs(cos(eta));
+                /* Fudge parameter to make sure, that TWTS pulse starts to impact simulation volume
+                 * at low intensity values. */
+                const float_64 m=3.;
+                /* Approximate cross section of laser pulse through y-axis,
+                 * scaled with "fudge factor" m. */
+                const float_64 y2=m*(pulselength_SI*picongpu::SI::SPEED_OF_LIGHT_SI)/cos(eta);
+                /* y-position of laser coordinate system origin within simulation. */
+                const float_64 y3=focus_y_SI;
+                /* Programmatically obtained time-delay */
+                const float_64 tdelay= (y1+y2+y3)/(picongpu::SI::SPEED_OF_LIGHT_SI*beta_0);
+                
+                return time_SI-tdelay;
+            }
+            else
+                return time_SI-tdelay_user_SI;
+        }
+        
+        template<>
+        HDINLINE float3_X
+        TWTSFieldE::getTWTSEfield_Normalized<DIM3>(
+                    const PMacc::math::Vector<float3_64,FieldE::numComponents>& eFieldPositions_SI,
+                    const float_64 time) const
+        {
+            return float3_X( float_X( calcTWTSEx(eFieldPositions_SI[0],time) ),
+                             float_X(0.), float_X(0.) );
+        }
+        
+        template<>
+        HDINLINE float3_X
+        TWTSFieldE::getTWTSEfield_Normalized<DIM2>(
+            const PMacc::math::Vector<float3_64,FieldE::numComponents>& eFieldPositions_SI,
+            const float_64 time) const
+        {
+            /* Ex->Ez, so also the grid cell offset for Ez has to be used. */
+            return float3_X( float_X(0.), float_X(0.),
+                             float_X( calcTWTSEx(eFieldPositions_SI[2],time) ) );
+        }
+        
+        HDINLINE float3_X
+        TWTSFieldE::operator()( const DataSpace<simDim>& cellIdx,
+                                const uint32_t currentStep ) const
+        {
+            const float_64 time_SI=getTime_SI<simDim>(currentStep);
+            const PMacc::math::Vector<float3_64,FieldE::numComponents> eFieldPositions_SI=
+                                                            getEfieldPositions_SI<simDim>(cellIdx);
+            /* Single TWTS-Pulse */
+            return getTWTSEfield_Normalized<simDim>(eFieldPositions_SI, time_SI);
+        }
+
+        /** Calculate the Ex(r,t) field here
+         *
+         * \param pos Spatial position of the target field.
+         * \param time Absolute time (SI, including all offsets and transformations) for calculating
+         *             the field */
+        HDINLINE TWTSFieldE::float_T
+        TWTSFieldE::calcTWTSEx( const float3_64& pos, const float_64 time) const
```

Comment:
This one is different, because it always takes a 3D vector. The dimensional specialization is done in `getTWTSEfield_Normalized<simDim>()`. This keeps the core functions clean from clutter.